src/ILCompiler.Compiler/src/Compiler/ObjectDumper.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Xml;
+
+using Internal.Text;
+
+using ILCompiler.DependencyAnalysis;
+
+using ObjectData = ILCompiler.DependencyAnalysis.ObjectNode.ObjectData;
+
+namespace ILCompiler
+{
+    public class ObjectDumper : IObjectDumper
+    {
+        private readonly string _fileName;
+
+        private XmlWriter _writer;
+
+        public ObjectDumper(string fileName)
+        {
+            _fileName = fileName;
+        }
+
+        internal void Begin()
+        {
+            var settings = new XmlWriterSettings
+            {
+                CloseOutput = true,
+                Indent = true,
+            };
+
+            _writer = XmlWriter.Create(File.CreateText(_fileName), settings);
+            _writer.WriteStartElement("ObjectNodes");
+        }
+
+        void IObjectDumper.DumpObjectNode(NameMangler mangler, ObjectNode node, ObjectData objectData)
+        {
+            string name = null;
+
+            _writer.WriteStartElement(node.GetType().Name.Replace('`', '_'));
```

Comment:
Now that I think about it - how about I make it strip everything after (and including) the grave accent? Also, strip the `Node` suffix.

The log looks like this right now:

```xml
<ObjectNodes>
  <ReadyToRunHeaderNode Name="__ReadyToRunHeader" Length="616" />
  <ModulesSectionNode Name="__Module" Length="8" />
  <ArrayOfEmbeddedPointersNode_1 Length="3592" />
</ObjectNodes>
```

Might look better this way:

```xml
<ObjectNodes>
  <ReadyToRunHeader Name="__ReadyToRunHeader" Length="616" />
  <ModulesSection Name="__Module" Length="8" />
  <ArrayOfEmbeddedPointers Length="3592" />
</ObjectNodes>
```